src/nbl/video/IOpenGL_LogicalDevice.h

```diff
@@ -0,0 +1,86 @@
+#ifndef __NBL_I_OPENGL__LOGICAL_DEVICE_H_INCLUDED__
+#define __NBL_I_OPENGL__LOGICAL_DEVICE_H_INCLUDED__
+
+#include "nbl/video/ILogicalDevice.h"
+#include "nbl/video/CEGL.h"
+#include "nbl/system/IThreadHandler.h"
+#include "nbl/core/alloc/PoolAddressAllocator.h"
+
+namespace nbl {
+namespace video
+{
+
+// Common base for GL and GLES logical devices
+// All COpenGL* objects (buffers, images, views...) will keep pointer of this type (just to be able to request destruction in destructor though)
+// Implementation of both GL and GLES is the same code (see COpenGL_LogicalDevice) thanks to IOpenGL_FunctionTable abstraction layer
+class IOpenGL_LogicalDevice : public ILogicalDevice
+{
+protected:
+    enum E_REQUEST_TYPE
+    {
+        ERT_BUFFER_CREATE,
+        ERT_BUFFER_DESTROY,
+
+        ERT_BUFFER_VIEW_CREATE,
+        ERT_BUFFER_VIEW_DESTROY,
+
+        ERT_IMAGE_CREATE,
+        ERT_IMAGE_DESTROY,
+
+        ERT_IMAGE_VIEW_CREATE,
+        ERT_IMAGE_VIEW_DESTROY,
+
+        ERT_FRAMEBUFFER_CREATE,
+        ERT_FRAMEBUFFER_DESTROY,
+
+        ERT_SWAPCHAIN_CREATE,
+        ERT_SWAPCHAIN_DESTROY,
+
+        //....
+    };
+
+    struct SRequest
+    {
+        E_REQUEST_TYPE type;
+        // TODO uncomment below later
+        //std::variant<All,Possible,Types,Of,Parameter,Structs,For,Requests>
+
+        // only relevant in case of CREATE requests
+        core::smart_refctd_ptr<core::IReferenceCounted> retval;
```

Comment:
well you could also pump the pointer of "the object to be destroyed" there

Reply:
btw Vulkan is an evil bastard and we will need to keep an object alive as long as the implementation might be using it, so it means the refcounting will have to be extreme and the object will have to delete itself from the `drop`

Which means that uses of objects will need to refcount them, and not let go of the reference until the GPU finishes execution... but thats for the Vulkan backend/